src/types/proof/definition.rs

```diff
@@ -172,16 +186,188 @@ impl<H: SimpleHasher> SparseMerkleProof<H> {
+
+            match self.leaf {
+                // In the case there is a leaf in the Merkle path, we check that this leaf exists in the tree
+                // The inserted key is going to update an existing leaf
+                Some(leaf_node) => {
+                    if new_element_key == leaf_node.key_hash {
+                        // Case 1: The new element key is the same as the leaf key (value update)
+                        // Step 1: we verify the key (with the old value) is present in the Merkle tree
+                        self.verify_existence(
+                            old_root_hash,
+                            new_element_key,
+                            leaf_node.value_hash.0,
+                        )?;
+
+                        // Step 2: we compute the new Merkle path (we build a new [`SparseMerkleProof`] object)
+                        // In this case the siblings are left unchanged, only the leaf value is updated
+                        let new_merkle_path: SparseMerkleProof<H> = SparseMerkleProof::new(
+                            Some(SparseMerkleLeafNode::new(
+                                new_element_key,
+                                ValueHash::with::<H>(new_element_value),
+                            )),
+                            self.siblings,
+                        );
+
+                        // Step 3: we compare the new Merkle path against the new_root_hash
+                        new_merkle_path.unsafe_verify(new_root_hash, new_element_key)?;
+                    } else {
+                        // Case 2: The new element key is different from the leaf key (leaf creation)
+                        // Step 1: we verify the old key is going to be split following the insertion of the
+                        // new key (nonexistence proof)
+                        self.verify_nonexistence(old_root_hash, new_element_key)?;
+
+                        // Add the correct siblings of the new element key by finding the splitting nibble and
+                        // adding the default leaves to the path
+
+                        // To compute the number of default leaves we need to add, we need to:
+                        // - Compute the number of default leaves to separate the old leaf from the new leaf in the last nibble
+                        // - Compute the number of default leaves to traverse the common prefix
+                        // - Compute the number of default leaves remaining to select the former old leaf in the former last nibble
+                        // (this leaf becomes an internal node, hence the path needs to be fully specified)
+
+                        let new_key_path = NibblePath::new(new_element_key.0.to_vec());
+                        let old_key_path = NibblePath::new(leaf_node.key_hash.0.to_vec());
+
+                        // The verify_nonexistence check from before ensure that the common prefix nibbles_len is greater than the
```

Comment:
This logic is complicated enough that it probably warrants its own function. That way, we can also test it independently of the other update logic.